awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -45,7 +45,7 @@ class TowerModule(AnsibleModule):
     authenticated = False
     config_name = 'tower_cli.cfg'
 
-    def __init__(self, argument_spec, **kwargs):
+    def __init__(self, argument_spec, mock_params=None, **kwargs):
```

Comment:
I like the mocking of the input but how do the module.exit_json work with inventory and lookup plugins? Does whatever format the provide work properly with other plugin types?

Reply:
This is a problem with this PR as it stands.

I have a pretty clear idea of how it would be resolved, I just haven't had the time. The solution would almost certainly be an error callback method. This would be defined in the inventory plugin file and passed to the init method, along side of this. Obviously, it makes sense for the default behavior error behavior to be what the modules need, `exit_json`.

That is nice, because it only has to be written once, and the ansible parse error is imported in the right case (the same exception that was already used in the inventory plugin).

The error callback needs a stable contract. It could take a message and a dictionary of "extra" information (like the response text and code). Then the inventory plugin would be free to ignore the extra data. Alternatively, it could take a dict with a "msg" key or something like that, which it could count on always being present.

Reply:
We probably also need to think about the logout in there. Since we overloaded the existing fail/exit_json from the module we can be sure that the logout process is invoked even if someone were to call it randomly from within one of the module.

Reply:
sure, that requires some thought.

`make_request` wouldn't be calling `self.fail_json`, but some `self.proxy_fail`.

I don't know _exactly_ how that should be handled, because there's a hazard in that some modules call `fail_json` directly, in which case the logout method needs to be called. So then you're probably left calling the logout method before calling the error callback so that `fail_json` / `exit_json` can retain their current code structure.